apm-sniffer/apm-sdk-plugin/rocketMQ-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v4/OnSuccessInterceptor.java

```diff
@@ -42,14 +42,22 @@ public class OnSuccessInterceptor implements InstanceMethodsAroundInterceptor {
     public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
         MethodInterceptResult result) throws Throwable {
         SendCallBackEnhanceInfo enhanceInfo = (SendCallBackEnhanceInfo) objInst.getSkyWalkingDynamicField();
-        AbstractSpan activeSpan = ContextManager.createLocalSpan(CALLBACK_OPERATION_NAME_PREFIX + enhanceInfo.getTopicId() + "/Producer/Callback");
+        String topicId = MessageSendInterceptor.DEFAULT_TOPIC;
+        // The SendCallBackEnhanceInfo could be null when there is an internal exception in the client API,
+        // such as MQClientException("no route info of this topic")
```

Comment:
@nisiyong Any explanation? Could you provide demo codes to explain this?

Reply:
This comment is from `OnExceptionInterceptor`, I think it should be updated.

> RocketMQ is still calling `onSuccess` when `no route info of this topic` exception?

I am not sure about that, but the NPE still happens when calling `onSuccess`. And I think when the `allArguments[6]` is null in `MessageSendInterceptor`, it won't set the SkyWalkingDynamicField, then the NPE happens.


```java
package org.apache.skywalking.apm.plugin.rocketMQ.v4;

public class MessageSendInterceptor implements InstanceMethodsAroundInterceptor {

    ...

    @Override
    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
        MethodInterceptResult result) throws Throwable {
        ...

        if (allArguments[6] != null) {
            ((EnhancedInstance) allArguments[6]).setSkyWalkingDynamicField(new SendCallBackEnhanceInfo(message.getTopic(), ContextManager
                .capture()));
        }
    }

```

Reply:
What is `allArguments[6]`?

Reply:
It is the `SendCallback sendCallback` in [MQClientAPIImpl.java](https://github.com/apache/rocketmq/blob/release-4.3.0/client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java). (rocketmq-client-4.3.0)

<img width="636" alt="image" src="https://user-images.githubusercontent.com/8198862/168207204-9860a0d4-2db6-4df9-8203-c05504499271.png">

Reply:
I am confused about this case. How this callback is null, but `onSuccess` is being called? I could get the idea `onException` invoked.

Reply:
I think it is because the SendCallback works this with the async mode in RocketMQ. And the sync mode will pass the NULL SendCallback.

<img width="1647" alt="image" src="https://user-images.githubusercontent.com/8198862/168235042-59a93700-2532-4c6f-bb56-e750e0275d43.png">

* [DefaultMQProducerImpl.java](https://github.com/apache/rocketmq/blob/release-4.3.0/client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java)
<img width="858" alt="image" src="https://user-images.githubusercontent.com/8198862/168235257-d96329c9-089c-4e3e-a1e5-13b7d2e0d9ed.png">

Reply:
@nisiyong if the callback is null, there will be no ```onSuccess``` invoked.
recommended to provide code that can reproduce the problem.
Is it possible that it was not intercepted by ```MessageSendInterceptor```?

Reply:
Thanks, I will try to reproduce it.